agents/tektronix3021c/tektronix_agent.py

```diff
@@ -0,0 +1,146 @@
+
+        self.job = None
+        self.ip_address = ip_address
+        self.gpib_slot = gpib_slot
+        self.monitor = False
+
+        self.awg = None
+        # Registers data feeds
+        agg_params = {
+            'frame_length': 60,
+        }
+        self.agent.register_feed('AWG',
+                                 record=True,
+                                 agg_params=agg_params)
+
+
+    def init_awg(self, session, params=None):
+        """ Task to connect to Tektronix AWG """
+
+        with self.lock.acquire_timeout(0) as acquired:
+            if not acquired:
+                return False, "Could not acquire lock"
+
+            try:
+                self.awg = tektronixInterface(self.ip_address, self.gpib_slot)
+                self.idn = self.awg.identify()
+            except socket.timeout as e:
+                self.log.error("Tektronix AWG timed out during connect")
+                return False, "Timeout"
+            self.log.info("Connected to AWG: {}".format(self.idn))
+
+        return True, 'Initialized AWG.'
+
+
+    def set_frequency(self, session, params=None):
+        """
+        Sets frequency of function generator:
+
+        Args:
+            frequency (float): Frequency to set. Must be between 0 and 25,000,000
+        """
+
+        with self.lock.acquire_timeout(1) as acquired:
+            if acquired:
+                freq = params['frequency']
+                self.awg.setFreq(freq)
+                
+                data = {'timestamp': time.time(),
+                       'block_name': "AWG_frequency",
+                        'data': {'AWG_frequency': freq}
+                       }
+                self.agent.publish_to_feed('AWG', data)
+                
+            else:
+                return False, "Could not acquire lock"
+
+        return True, 'Set frequency {}'.format(params)
+
+    def set_amplitude(self, session, params=None):
+        """
+        Sets current of power supply:
+
+        Args:
+            Amplitude (float): Peak to Peak voltage to set. Must be between 0 and 10.
+        """
+        with self.lock.acquire_timeout(1) as acquired:
+            if acquired:
+                amp = params['amplitude']
+                self.awg.setAmp(amp)
+                
+                data = {'timestamp': time.time(),
+                       'block_name': "AWG_amplitude",
+                        'data': {'AWG_amplitude': amp}
+                       }
+                self.agent.publish_to_feed('AWG', data)
+                
+            else:
+                return False, "Could not acquire lock"
+
+        return True, 'Set amplitude to {} '.format(params)
+
+    def set_output(self, session, params=None):
+        """
+        Task to turn channel on or off.
+
+        Args:
+            state (bool): True for on, False for off.
+        """
+        with self.lock.acquire_timeout(1) as acquired:
+            if acquired:
+                state = params['state']
```

Comment:
It would be safer to use a `params.get('state')` with a default value or perhaps `None` and some error handling for if a user doesn't pass in a valid value (or no value at all) to avoid a `KeyError` or the like.